src/workerd/io/worker.c++

```diff
@@ -1850,6 +1863,39 @@ Worker::Worker(kj::Own<const Script> scriptParam,
   });
 }
 
+void Worker::runInUnloadScope(IoContext& ctx, kj::Function<void(jsg::Lock&)> func) const {
+  auto workerWeakRef = kj::atomicAddRefWeak(*this);
+  ctx.addWaitUntil(takeAsyncLockWithoutRequest(nullptr).then(
+                       [workerWeakRef = kj::mv(workerWeakRef), func = kj::mv(func)](
+                           AsyncLock asyncLock) mutable -> kj::Promise<void> {
+    // If worker is already disposed, no need to run unload handler.
+    KJ_IF_SOME(worker, workerWeakRef) {
+      worker->runInLockScope(asyncLock, [&](Lock& lock) {
+        JSG_WITHIN_CONTEXT_SCOPE(lock, lock.getContext(), [&](jsg::Lock& js) {
```

Comment:
We need to be setting up a LimitEnforcer somewhere here, otherwise unload events will be allowed to go into infinite loops. I suppose we should probably try to take ownership of the `LimitEnforcer` from the `IoContext`, so that we continue executing under the same CPU limit?

Generally we need to look carefully at everything `IoContext::run()` sets up and do similar things here. Or alternatively look at what the constructor of `Worker` or `Worker::Script` sets up.